### PR TITLE
refactor(Compiler): inline styles before compiling the template

### PR DESCRIPTION
**Pull request by vicb**
TODO:
- [x] ~~add ref to 1694 to the commit message~~ - fixed in a previous PR
- [x] ~~implement relative url resolving to fix JS~~
- [x] find a way to support `querySelectorAll` in Dart
- [x] address TODOs (1 xited tests)

### CODE REVIEW

File: modules/angular2/src/render/dom/compiler/compiler.ts [hunk 1]
```diff
@@ -54,17 +54,13 @@ export class DomCompiler extends RenderCompiler {
...
     var compileElements = pipeline.process(tplElement, protoViewType, viewDef.componentId);
 
     var protoView = compileElements[0].inheritedProtoView.build(this._propertySetterFactory);
 
-    if (subTaskPromises.length > 0) {
-      return PromiseWrapper.all(subTaskPromises).then((_) => protoView);
-    } else {
-      return PromiseWrapper.resolve(protoView);
-    }
+    // TODO(vicb): remove Promise
```
Comment (tbosch) [inline modules/angular2/src/render/dom/compiler/compiler.ts]: What is the blocker here?

File: modules/angular2/src/render/dom/compiler/template_loader.ts [hunk 2]
```diff
@@ -67,10 +68,75 @@ export class TemplateLoader {
...
+    if (isPresent(view.template)) {
+      html = PromiseWrapper.resolve(view.template);
+    } else if (isPresent(view.templateAbsUrl)) {
+      html = this._loadText(view.templateAbsUrl);
+    } else {
+      throw new BaseException('View should have either the templateUrl or template property set');
+    }
+
+    // Inline the style tags from the html
+    return html.then(html => {
```
Comment (tbosch) [inline modules/angular2/src/render/dom/compiler/template_loader.ts]: Maybe factor out into a separate method?

File: modules/angular2/src/render/dom/shadow_dom/shadow_dom_strategy.ts [hunk 3]
```diff
@@ -5,24 +5,17 @@ import * as viewModule from '../view/view';
 import {LightDom} from './light_dom';
 
 export class ShadowDomStrategy {
+  // Whether the stratgy understands the native <content> tag
```
Comment (tbosch) [inline modules/angular2/src/render/dom/shadow_dom/shadow_dom_strategy.ts]: Why change the comment from `/**/` to `//`? AFAIK, `/**/` will be used to generate docs, but `//` will not be used.

Comment (tbosch) [inline modules/angular2/src/render/dom/shadow_dom/shadow_dom_strategy.ts]: @naomiblack @petebacondarwin could you explain what kind of comments end up in our docs and which ones don't?

**Comment by tbosch** [comment]
@vicb This looks good! Is it still WIP or besides the Travis errors good to go?

File: modules/angular2/src/render/dom/compiler/compiler.ts (hunk 1 above)
Comment (vicb) [inline modules/angular2/src/render/dom/compiler/compiler.ts]: no blocker, I'm on it

File: modules/angular2/src/render/dom/compiler/template_loader.ts (hunk 2 above)
Comment (vicb) [inline modules/angular2/src/render/dom/compiler/template_loader.ts]: I don't know... the code is used only once and:
- I can't think of a good method name,
- The code has little sense outside the context of the enclosing method.

**Comment by vicb** [comment]
@tbosch I have currently to issues with this PR:
(/ref https://travis-ci.org/angular/angular/builds/67068776)

**in Dart**

We use the `Html5LibDomAdapter` adapter and it does not support `querySelectorAll` which is needed to find `<style>` tags in the template.

Any idea or an idea on who can help ?

**in JS**

Because relative url resolution is not (yet) working, we fail to resolve css urls (in material examples). It throws an exception and tests fail.

The best for JS is probably to wait for the resolution to be implemented - I'll start working on this tomorrow.

**Comment by tbosch** [comment]
@vicb: Could you walk the template manually, instead of using `querySelectorAll`?

Regarding JS: How did it work before? I.e. we were able to resolve the url for the html file correctly... Don't see the connection to relative url resolution...

**Comment by vicb** [comment]
@tbosch 
- for JS we used not to inline `@import` rules in emulated unscoped which explain the current issue,
- for Dart I would prefer not to reinvent the wheel.

**Comment by vicb** [comment]
Blocked on the Dart issue describe in the comments above, any help would be appreciated

**Comment by vicb** [comment]
I got some hints from @yjbanov (via @tbosch) to help unblocking this issue, I'll try to implement it next week.

**Comment by vicb** [comment]
@tbosch could you please review the PR ?

Highlights:
- UrlResolver is no more relying on a anchor element. It can now be used in any env (ie server, WebWorkers),
- Styles get inlined before the template is actually compiled,
- The JS server DOM adapter has been fixed so that it doesn't try to insert empty text nodes,
- Last commits are all bout cleanup

I have kept the changes in separate commits in order to make the review less painful.

**Comment by tbosch** [comment]
Do you know how and whether the transformer also inlines styles / templates? Do they use our `TemplateLoader` as well for this?

**Comment by tbosch** [comment]
LGTM.
Could you create a new issue for replacing your `UrlResolver` with a third party library and please ping Igor and Brian on that issue as well for their url parsing needs.

**Comment by vicb** [comment]
> Do you know how and whether the transformer also inlines styles / templates? Do they use our TemplateLoader as well for this?

I think it has been deferred, see https://github.com/angular/angular/pull/2599#issuecomment-113332884 (2 new issues have been created to track that)

**Comment by vicb** [comment]
landed as a0e0f3123b0e65c5d354614734d9965031cab196 & previous commits.

Thanks for the review @tbosch

**Comment by yjbanov** [comment]
Just a reminder that renames should be marked as "BREAKING CHANGES" so our changelog could pick them up.

**Pull request closed without merging**